.circleci/config.yml

```diff
@@ -1,109 +1,303 @@
-version: 2
-jobs:
-  build:
-    parallelism: 6
-    working_directory: ~/wp-calypso
-    docker:
-      - image: circleci/node:10.10.0-browsers
-        environment:
-          CIRCLE_ARTIFACTS: /tmp/artifacts
-          CIRCLE_TEST_REPORTS: /tmp/test_results
+references:
+  shared-environment: &shared-environment
+    CIRCLE_ARTIFACTS: /tmp/artifacts
+    CIRCLE_TEST_REPORTS: /tmp/test_results
 
-    steps:
-      - restore_cache:
-          name: Restore git cache
-          key: v1-20180524-git
-
-      - checkout
-
-    # Update origin/master so we can diff against it later
-      - run:
-          name: Update master branch
-          command: git fetch --force origin master
+  docker-image: &docker-image
+    - image: circleci/node:10.10.0-browsers
 
-      - save_cache:
-          name: Save git cache
-          key: v1-20180524-git
-          paths:
-            - ".git"
+  setup-results-and-artifacts: &setup-results-and-artifacts
+    name: Create Directories for Results and Artifacts
+    command: |
+      mkdir -p                             \
+        "$CIRCLE_ARTIFACTS/translate"      \
+        "$CIRCLE_TEST_REPORTS/client"      \
+        "$CIRCLE_TEST_REPORTS/eslint"      \
+        "$CIRCLE_TEST_REPORTS/integration" \
+        "$CIRCLE_TEST_REPORTS/server"      \
+        "$HOME/jest-cache"
 
-      - run:
-          name: Create Directories for Results and Artifacts
-          command: |
-            mkdir -p $CIRCLE_ARTIFACTS
-            mkdir -p $CIRCLE_TEST_REPORTS
+  restore-jest-cache: &restore-jest-cache
+    name: Restore Jest cache
+    keys:
+      - v6-jest-{{ .Environment.CIRCLE_JOB }}-{{ .Environment.CIRCLE_NODE_INDEX }}/{{ .Environment.CIRCLE_NODE_TOTAL }}-{{ .Branch }}-{{ .Revision }}
+      - v6-jest-{{ .Environment.CIRCLE_JOB }}-{{ .Environment.CIRCLE_NODE_INDEX }}/{{ .Environment.CIRCLE_NODE_TOTAL }}-{{ .Branch }}
+      - v6-jest-{{ .Environment.CIRCLE_JOB }}-{{ .Environment.CIRCLE_NODE_INDEX }}/{{ .Environment.CIRCLE_NODE_TOTAL }}-master
+      - v6-jest-{{ .Environment.CIRCLE_JOB }}-{{ .Environment.CIRCLE_NODE_INDEX }}/{{ .Environment.CIRCLE_NODE_TOTAL }}
```

Comment:
I see we're storing Jest cache using the `-{{ .Branch }}-{{ .Revision }}` key below, but where are we using the others?

Reply:
These are the restores, so CircleCI will try to match from first to last until it gets a hit or fails and doesn't restore a cache.

- Ideally, we have a cache from this revision.
- Otherwise, from this branch
- Otherwise, from master
- …

---

The `v*` prefixes are just ways of cache busting. CircleCI caches do not necessarily expire. As I've iterated on this I've used several, which is how we got a relatively high number. Given that caches are never overwritten, if something needs to change in the way things are cached, it's a good idea to cache bust by bumping this.

Reply:
> Otherwise, from master

This isn't really master though. It's whatever master was the first time we had a run for master. I'm not sure it's really a good fallback...

Reply:
> This isn't really master though.

It _could_ be master, but mostly it won't 🙂.  master probably has a decent cache we can leverage rather than starting from scratch. We're just warming up the Jest cache, Jest will be able to leverage some of the cache from master, and what it doesn't use I wouldn't expect to be harmful.

> It's whatever master was the first time we had a run for master.

It should be from the HEAD, unless we somehow generate a cache key from an old commit on the master branch. Master caches will have keys like

```
v6-jest-test-client-0/5-master-abcd123
```

A new cache is created for each commit on each branch, and we default to the [most recent one](https://circleci.com/docs/2.0/caching/#restoring-cache):

> The cache will be restored from the first matching key. If there are multiple matches, the most recently generated cache will be used.

If PRs aren't "far" from master, this is probably a good way to warm the cache.

Does that seem like sound reasoning?

Reply:
ah, interesting, I missed the most-recent-cache bit on restores. Cool!